# Naive Bayes (Salary Data)

In [1]:
import numpy as np
import pandas as pd
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

In [2]:
import pandas as pd
X_train = pd.read_csv('SalaryData_Train.csv',encoding= 'ISO-8859-1')
X_test= pd.read_csv("SalaryData_Test.csv",encoding = 'ISO-8859-1')

In [3]:
df = pd.concat([X_train,X_test])

# Explore Pre-Processing

In [4]:
df.shape

(45221, 14)

In [5]:
df1 = df.copy()

In [6]:
df1.shape

(45221, 14)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45221 entries, 0 to 15059
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            45221 non-null  int64 
 1   workclass      45221 non-null  object
 2   education      45221 non-null  object
 3   educationno    45221 non-null  int64 
 4   maritalstatus  45221 non-null  object
 5   occupation     45221 non-null  object
 6   relationship   45221 non-null  object
 7   race           45221 non-null  object
 8   sex            45221 non-null  object
 9   capitalgain    45221 non-null  int64 
 10  capitalloss    45221 non-null  int64 
 11  hoursperweek   45221 non-null  int64 
 12  native         45221 non-null  object
 13  Salary         45221 non-null  object
dtypes: int64(5), object(9)
memory usage: 5.2+ MB


In [8]:
df.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            13   United-States   <=50K  
2            0            40   United-States   <=50K  
3            0            40   United-States   <=50K  
4            0            40            Cuba   <=50K

In [9]:
df.education.unique()

array([' Bachelors', ' HS-grad', ' 11th', ' Masters', ' 9th',
       ' Some-college', ' Assoc-acdm', ' 7th-8th', ' Doctorate',
       ' Assoc-voc', ' Prof-school', ' 5th-6th', ' 10th', ' Preschool',
       ' 12th', ' 1st-4th'], dtype=object)

In [10]:
df.occupation.unique()

array([' Adm-clerical', ' Exec-managerial', ' Handlers-cleaners',
       ' Prof-specialty', ' Other-service', ' Sales', ' Transport-moving',
       ' Farming-fishing', ' Machine-op-inspct', ' Tech-support',
       ' Craft-repair', ' Protective-serv', ' Armed-Forces',
       ' Priv-house-serv'], dtype=object)

In [11]:
#Above col have more unique variable then it will form more number of column so we apply encoding technique on it
#and extract some features

In [12]:
# display categorical variables
categorical = [col for col in X_train.columns if X_train[col].dtypes == 'O']
categorical

['workclass',
 'education',
 'maritalstatus',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native',
 'Salary']

In [13]:
pip install --upgrade category_encoders

Note: you may need to restart the kernel to use updated packages.


In [14]:
import category_encoders as ce

In [15]:
# encode remaining variables with one-hot encoding
encoder = ce.OneHotEncoder(cols=['workclass',
 'education',
 'maritalstatus',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native'])
df = encoder.fit_transform(df)

In [16]:
df

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
0       39            1            0            0            0            0   
1       50            0            1            0            0            0   
2       38            0            0            1            0            0   
3       53            0            0            1            0            0   
4       28            0            0            1            0            0   
...    ...          ...          ...          ...          ...          ...   
15055   33            0            0            1            0            0   
15056   39            0            0            1            0            0   
15057   38            0            0            1            0            0   
15058   44            0            0            1            0            0   
15059   35            0            0            0            0            0   

       workclass_6  workclass_7  education_1  education_2  ...  native_32  \
0                0            0            1            0  ...          0   
1                0            0            1            0  ...          0   
2                0            0            0            1  ...          0   
3                0            0            0            0  ...          0   
4                0            0            1            0  ...          0   
...            ...          ...          ...          ...  ...        ...   
15055            0            0            1            0  ...          0   
15056            0            0            1            0  ...          0   
15057            0            0            1            0  ...          0   
15058            0            0            1            0  ...          0   
15059            1            0            1            0  ...          0   

       native_33  native_34  native_35  native_36  native_37  native_38  \
0              0          0          0          0          0          0   
1              0          0          0          0          0          0   
2              0          0          0          0          0          0   
3              0          0          0          0          0          0   
4              0          0          0          0          0          0   
...          ...        ...        ...        ...        ...        ...   
15055          0          0          0          0          0          0   
15056          0          0          0          0          0          0   
15057          0          0          0          0          0          0   
15058          0          0          0          0          0          0   
15059          0          0          0          0          0          0   

       native_39  native_40  Salary  
0              0          0   <=50K  
1              0          0   <=50K  
2              0          0   <=50K  
3              0          0   <=50K  
4              0          0   <=50K  
...          ...        ...     ...  
15055          0          0   <=50K  
15056          0          0   <=50K  
15057          0          0   <=50K  
15058          0          0   <=50K  
15059          0          0    >50K  

[45221 rows x 103 columns]

In [17]:
X=df.drop(['Salary'], axis=1)
y =df['Salary']

In [18]:
#apply SelectKBest class to extract top 10 best features
bestfeatures = SelectKBest(score_func=chi2, k=20)
fit = bestfeatures.fit(X,y)
dfscores = pd.DataFrame(fit.scores_)
dfcolumns = pd.DataFrame(X.columns)
#concat two dataframes for better visualization 
featureScores = pd.concat([dfcolumns,dfscores],axis=1)
featureScores.columns = ['Specs','Score']
print(featureScores.nlargest(20,'Score'))

              Specs         Score
59      capitalgain  1.130196e+08
60      capitalloss  1.853151e+06
0               age  1.151555e+04
61     hoursperweek  8.221133e+03
26  maritalstatus_2  4.811024e+03
47   relationship_2  4.327760e+03
24      educationno  3.225929e+03
25  maritalstatus_1  3.126615e+03
49   relationship_4  1.913133e+03
33     occupation_2  1.716744e+03
58            sex_2  1.420945e+03
35     occupation_4  1.317871e+03
46   relationship_1  1.282016e+03
11      education_4  1.264789e+03
8       education_1  1.200843e+03
36     occupation_5  1.106025e+03
18     education_11  1.079386e+03
50   relationship_5  8.768964e+02
6       workclass_6  8.279557e+02
27  maritalstatus_3  6.987953e+02


In [19]:
#As we see in top 20 these are the best feature from data base after one hot encoding 
#So wasting of analysis time we drop the unnecessary column are they:
#education already labelled feature
#race column
#native

In [20]:
#Again we read the data drop those column and apply one hot encoding.

In [21]:
df1=df1.drop(['education','race','native'],axis =1)

In [22]:
df1

age          workclass  educationno        maritalstatus  \
0       39          State-gov           13        Never-married   
1       50   Self-emp-not-inc           13   Married-civ-spouse   
2       38            Private            9             Divorced   
3       53            Private            7   Married-civ-spouse   
4       28            Private           13   Married-civ-spouse   
...    ...                ...          ...                  ...   
15055   33            Private           13        Never-married   
15056   39            Private           13             Divorced   
15057   38            Private           13   Married-civ-spouse   
15058   44            Private           13             Divorced   
15059   35       Self-emp-inc           13   Married-civ-spouse   

               occupation    relationship      sex  capitalgain  capitalloss  \
0            Adm-clerical   Not-in-family     Male         2174            0   
1         Exec-managerial         Husband     Male            0            0   
2       Handlers-cleaners   Not-in-family     Male            0            0   
3       Handlers-cleaners         Husband     Male            0            0   
4          Prof-specialty            Wife   Female            0            0   
...                   ...             ...      ...          ...          ...   
15055      Prof-specialty       Own-child     Male            0            0   
15056      Prof-specialty   Not-in-family   Female            0            0   
15057      Prof-specialty         Husband     Male            0            0   
15058        Adm-clerical       Own-child     Male         5455            0   
15059     Exec-managerial         Husband     Male            0            0   

       hoursperweek  Salary  
0                40   <=50K  
1                13   <=50K  
2                40   <=50K  
3                40   <=50K  
4                40   <=50K  
...             ...     ...  
15055            40   <=50K  
15056            36   <=50K  
15057            50   <=50K  
15058            40   <=50K  
15059            60    >50K  

[45221 rows x 11 columns]

In [23]:
#Remove those unnecessary 3 column

In [24]:
# display categorical variables
categorical = [col for col in df1.columns if df1[col].dtypes == 'O']
categorical

['workclass', 'maritalstatus', 'occupation', 'relationship', 'sex', 'Salary']

In [25]:
#Here workclass is a ordinal column so we used label encoding for workclassb
from sklearn import preprocessing
label = preprocessing.LabelEncoder()

In [26]:
df1['workclass']= label.fit_transform(df1['workclass'])
df1['workclass'].unique()

array([5, 4, 2, 0, 1, 3, 6])

In [27]:
#apply OneHotEncoding

In [28]:
# encode remaining variables with one-hot encoding
encoder = ce.OneHotEncoder(cols=['maritalstatus', 'occupation', 'relationship', 'sex'])
df1 = encoder.fit_transform(df1)

In [29]:
df1

age  workclass  educationno  maritalstatus_1  maritalstatus_2  \
0       39          5           13                1                0   
1       50          4           13                0                1   
2       38          2            9                0                0   
3       53          2            7                0                1   
4       28          2           13                0                1   
...    ...        ...          ...              ...              ...   
15055   33          2           13                1                0   
15056   39          2           13                0                0   
15057   38          2           13                0                1   
15058   44          2           13                0                0   
15059   35          3           13                0                1   

       maritalstatus_3  maritalstatus_4  maritalstatus_5  maritalstatus_6  \
0                    0                0                0                0   
1                    0                0                0                0   
2                    1                0                0                0   
3                    0                0                0                0   
4                    0                0                0                0   
...                ...              ...              ...              ...   
15055                0                0                0                0   
15056                1                0                0                0   
15057                0                0                0                0   
15058                1                0                0                0   
15059                0                0                0                0   

       maritalstatus_7  ...  relationship_3  relationship_4  relationship_5  \
0                    0  ...               0               0               0   
1                    0  ...               0               0               0   
2                    0  ...               0               0               0   
3                    0  ...               0               0               0   
4                    0  ...               1               0               0   
...                ...  ...             ...             ...             ...   
15055                0  ...               0               1               0   
15056                0  ...               0               0               0   
15057                0  ...               0               0               0   
15058                0  ...               0               1               0   
15059                0  ...               0               0               0   

       relationship_6  sex_1  sex_2  capitalgain  capitalloss  hoursperweek  \
0                   0      1      0         2174            0            40   
1                   0      1      0            0            0            13   
2                   0      1      0            0            0            40   
3                   0      1      0            0            0            40   
4                   0      0      1            0            0            40   
...               ...    ...    ...          ...          ...           ...   
15055               0      1      0            0            0            40   
15056               0      0      1            0            0            36   
15057               0      1      0            0            0            50   
15058               0      1      0         5455            0            40   
15059               0      1      0            0            0            60   

       Salary  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
15055   <=50K  
15056   <=50K  
15057   <=50K  
15058   <=50K  
15059    >50K  

[45221 rows x 36 columns]

In [30]:
X = df1.drop(['Salary'],axis = 1)
y = df1['Salary']

In [31]:
#To get top feature apply KBestFeature method
#apply SelectKBest class to extract top 12 best features
bestfeatures = SelectKBest(score_func=chi2, k=12)
fit = bestfeatures.fit(X,y)
dfscores = pd.DataFrame(fit.scores_)
dfcolumns = pd.DataFrame(X.columns)
#concat two dataframes for better visualization 
featureScores = pd.concat([dfcolumns,dfscores],axis=1)
featureScores.columns = ['Specs','Score']  #naming the dataframe columns
print(featureScores.nlargest(12,'Score'))  #print 26 best features

              Specs         Score
32      capitalgain  1.130196e+08
33      capitalloss  1.853151e+06
0               age  1.151555e+04
34     hoursperweek  8.221133e+03
4   maritalstatus_2  4.811024e+03
25   relationship_2  4.327760e+03
2       educationno  3.225929e+03
3   maritalstatus_1  3.126615e+03
27   relationship_4  1.913133e+03
11     occupation_2  1.716744e+03
31            sex_2  1.420945e+03
13     occupation_4  1.317871e+03


# FEATURE SCALING

In [32]:
#we use robust scaler here to standardized the data it will transform the data into 1st and 

In [33]:
cols = X.columns

In [34]:
cols

Index(['age', 'workclass', 'educationno', 'maritalstatus_1', 'maritalstatus_2',
       'maritalstatus_3', 'maritalstatus_4', 'maritalstatus_5',
       'maritalstatus_6', 'maritalstatus_7', 'occupation_1', 'occupation_2',
       'occupation_3', 'occupation_4', 'occupation_5', 'occupation_6',
       'occupation_7', 'occupation_8', 'occupation_9', 'occupation_10',
       'occupation_11', 'occupation_12', 'occupation_13', 'occupation_14',
       'relationship_1', 'relationship_2', 'relationship_3', 'relationship_4',
       'relationship_5', 'relationship_6', 'sex_1', 'sex_2', 'capitalgain',
       'capitalloss', 'hoursperweek'],
      dtype='object')

# Apply Encoder

In [35]:
#Standardizing data to work better for model building
from sklearn.preprocessing import RobustScaler
scaler = RobustScaler()
X = scaler.fit_transform(X)
X = scaler.transform(X)

C:\Users\DELL\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but RobustScaler was fitted with feature names
  warnings.warn(


In [36]:
X= pd.DataFrame(X, columns=[cols])

In [37]:
X

age workclass educationno maritalstatus_1 maritalstatus_2  \
0     -1.941828       1.0     -2.3125             1.0             0.0   
1     -1.911357       0.0     -2.3125             0.0             1.0   
2     -1.944598      -2.0     -2.5625             0.0             0.0   
3     -1.903047      -2.0     -2.6875             0.0             1.0   
4     -1.972299      -2.0     -2.3125             0.0             1.0   
...         ...       ...         ...             ...             ...   
45216 -1.958449      -2.0     -2.3125             1.0             0.0   
45217 -1.941828      -2.0     -2.3125             0.0             0.0   
45218 -1.944598      -2.0     -2.3125             0.0             1.0   
45219 -1.927978      -2.0     -2.3125             0.0             0.0   
45220 -1.952909      -1.0     -2.3125             0.0             1.0   

      maritalstatus_3 maritalstatus_4 maritalstatus_5 maritalstatus_6  \
0                 0.0             0.0             0.0             0.0   
1                 0.0             0.0             0.0             0.0   
2                 1.0             0.0             0.0             0.0   
3                 0.0             0.0             0.0             0.0   
4                 0.0             0.0             0.0             0.0   
...               ...             ...             ...             ...   
45216             0.0             0.0             0.0             0.0   
45217             1.0             0.0             0.0             0.0   
45218             0.0             0.0             0.0             0.0   
45219             1.0             0.0             0.0             0.0   
45220             0.0             0.0             0.0             0.0   

      maritalstatus_7  ... relationship_2 relationship_3 relationship_4  \
0                 0.0  ...            0.0            0.0            0.0   
1                 0.0  ...            1.0            0.0            0.0   
2                 0.0  ...            0.0            0.0            0.0   
3                 0.0  ...            1.0            0.0            0.0   
4                 0.0  ...            0.0            1.0            0.0   
...               ...  ...            ...            ...            ...   
45216             0.0  ...            0.0            0.0            1.0   
45217             0.0  ...            0.0            0.0            0.0   
45218             0.0  ...            1.0            0.0            0.0   
45219             0.0  ...            0.0            0.0            1.0   
45220             0.0  ...            1.0            0.0            0.0   

      relationship_5 relationship_6 sex_1 sex_2 capitalgain capitalloss  \
0                0.0            0.0  -1.0   0.0      2174.0         0.0   
1                0.0            0.0  -1.0   0.0         0.0         0.0   
2                0.0            0.0  -1.0   0.0         0.0         0.0   
3                0.0            0.0  -1.0   0.0         0.0         0.0   
4                0.0            0.0  -2.0   1.0         0.0         0.0   
...              ...            ...   ...   ...         ...         ...   
45216            0.0            0.0  -1.0   0.0         0.0         0.0   
45217            0.0            0.0  -2.0   1.0         0.0         0.0   
45218            0.0            0.0  -1.0   0.0         0.0         0.0   
45219            0.0            0.0  -1.0   0.0      5455.0         0.0   
45220            0.0            0.0  -1.0   0.0         0.0         0.0   

      hoursperweek  
0            -8.00  
1            -9.08  
2            -8.00  
3            -8.00  
4            -8.00  
...            ...  
45216        -8.00  
45217        -8.16  
45218        -7.60  
45219        -8.00  
45220        -7.20  

[45221 rows x 35 columns]

In [38]:
final_X = X[['age', 'educationno', 'maritalstatus_1', 'maritalstatus_2',
        'occupation_2',
       'occupation_4',
       'relationship_2','relationship_4',
       'sex_2', 'capitalgain',
       'capitalloss', 'hoursperweek']]

In [39]:
final_X

age educationno maritalstatus_1 maritalstatus_2 occupation_2  \
0     -1.941828     -2.3125             1.0             0.0          0.0   
1     -1.911357     -2.3125             0.0             1.0          1.0   
2     -1.944598     -2.5625             0.0             0.0          0.0   
3     -1.903047     -2.6875             0.0             1.0          0.0   
4     -1.972299     -2.3125             0.0             1.0          0.0   
...         ...         ...             ...             ...          ...   
45216 -1.958449     -2.3125             1.0             0.0          0.0   
45217 -1.941828     -2.3125             0.0             0.0          0.0   
45218 -1.944598     -2.3125             0.0             1.0          0.0   
45219 -1.927978     -2.3125             0.0             0.0          0.0   
45220 -1.952909     -2.3125             0.0             1.0          1.0   

      occupation_4 relationship_2 relationship_4 sex_2 capitalgain  \
0              0.0            0.0            0.0   0.0      2174.0   
1              0.0            1.0            0.0   0.0         0.0   
2              0.0            0.0            0.0   0.0         0.0   
3              0.0            1.0            0.0   0.0         0.0   
4              1.0            0.0            0.0   1.0         0.0   
...            ...            ...            ...   ...         ...   
45216          1.0            0.0            1.0   0.0         0.0   
45217          1.0            0.0            0.0   1.0         0.0   
45218          1.0            1.0            0.0   0.0         0.0   
45219          0.0            0.0            1.0   0.0      5455.0   
45220          0.0            1.0            0.0   0.0         0.0   

      capitalloss hoursperweek  
0             0.0        -8.00  
1             0.0        -9.08  
2             0.0        -8.00  
3             0.0        -8.00  
4             0.0        -8.00  
...           ...          ...  
45216         0.0        -8.00  
45217         0.0        -8.16  
45218         0.0        -7.60  
45219         0.0        -8.00  
45220         0.0        -7.20  

[45221 rows x 12 columns]

In [40]:
y

0         <=50K
1         <=50K
2         <=50K
3         <=50K
4         <=50K
          ...  
15055     <=50K
15056     <=50K
15057     <=50K
15058     <=50K
15059      >50K
Name: Salary, Length: 45221, dtype: object

# Model BUilding

In [41]:
#split X and y into training and testing sets
from sklearn.model_selection import train_test_split

In [43]:
X_train, X_test, y_train, y_test = train_test_split(final_X, y, test_size = 0.3, random_state = 0)

In [44]:
# train a Gaussian Naive Bayes classifier on the training set
from sklearn.naive_bayes import GaussianNB

In [45]:
# instantiate the model
gnb = GaussianNB()

In [47]:
# fit the model
gnb.fit(X_train, y_train)   #gnb.fit(X_train,Y_train)

GaussianNB()

# Model Test

In [48]:
y_pred = gnb.predict(X_test)

In [49]:
y_pred

array([' <=50K', ' <=50K', ' <=50K', ..., ' <=50K', ' <=50K', ' <=50K'],
      dtype='<U6')

In [50]:
from sklearn.metrics import accuracy_score

print('Model accuracy score: {0:0.4f}'. format(accuracy_score(y_test, y_pred)))

Model accuracy score: 0.8238


# Compare the train-set and test-set accuracy

In [51]:
y_pred_train = gnb.predict(X_train)
y_pred_train

array([' <=50K', ' <=50K', ' <=50K', ..., ' >50K', ' <=50K', ' <=50K'],
      dtype='<U6')

In [52]:
from sklearn.metrics import accuracy_score
print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, y_pred_train)))

Training-set accuracy score: 0.8194


In [53]:
print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, y_pred_train)))

Training-set accuracy score: 0.8194


#### The training-set accuracy score is 0.8194 this value are quite comparable So there is no sign of overfitting.